### PR TITLE
Add "Processors": (generalize magicgui return type behavior outside of magicgui)

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
Another part of #4532, this begins to generalize the behavior used by magicgui that assigns "intentionality" to return type annotations.  It introduces "processors" which are functions that can be used to handle data of a given return type.  It would allow some function like this to both receive the currently selected layer, and then have its return value added as an Image layer to the viewer.

```py
@inject_napari_dependencies
def func(img: napari.types.ImageData) -> napari.types.ImageData:
    ...
```

(i.e. `@inject_napari_dependencies` now has a very similar effect to what you would have gotten with a `@magicgui` decorated function with the same return annotation)

## Type of change
- [x] New feature (non-breaking change which adds functionality)

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by tlambert03** [comment]
taking out of `[WIP]` since this is becoming a somewhat complete bit of work here (mostly just extracting from the `_magicgui` code what now can live as a more general "processor".

File: napari/utils/_injection/__init__.py
Comment (alisterburt) [inline napari/utils/_injection/__init__.py]: :)

File: napari/utils/_injection/__init__.py
Comment (alisterburt) [inline napari/utils/_injection/__init__.py:108]: :) again

File: napari/utils/_injection/__init__.py
Comment (alisterburt) [inline napari/utils/_injection/__init__.py]: super minor: found myself having to come back up here to figure out what `required` was, could consider refactoring to `required_parameters`?

File: napari/utils/_injection/_processors.py
Comment (alisterburt) [inline napari/utils/_injection/_processors.py:87]: I was a bit surprised by `Any` types here - is there a reason these shouldn't be more specific?

File: napari/utils/_injection/_processors.py
Comment (alisterburt) [inline napari/utils/_injection/_processors.py]: incomplete docstring here

File: napari/utils/_injection/_accessors.py
Comment (alisterburt) [inline napari/utils/_injection/_accessors.py]: super minor: slight preference for `set_accessors` given that the first argument is a dict

same goes for processors

**Review by alisterburt — COMMENTED** [review]
Seeing this come together is really awesome @tlambert03, I really like the `accessor`/`processor` patterns and their implementations/usage in this PR 🙂  have left a few minor comments - please ping me once you've looked at these minor things and if you feel like the PR is ready I'll spend some real time reviewing it over the weekend

File: napari/utils/_injection/__init__.py
Comment (tlambert03) [reply napari/utils/_injection/__init__.py:108]: everyone loves a good := 🦭

Comment (alisterburt) [reply napari/utils/_injection/__init__.py:108]: 🦭 is cool but mostly just feeling great about the clarity of the implementation :)

File: napari/utils/_injection/_processors.py
Comment (tlambert03) [reply napari/utils/_injection/_processors.py:87]: the `data` could be technically, but it's hard because we don't have a globally applicable "type that could be data for some layer".   It's not even `LayerDataProtocol` because of `Shapes`, I supposed you could theoretically come up with some Union of all the things that will possibly work to create data in at least one Layer type, but that seemed a bit out of scope for a private function :)

the `return_type`, should probably be `Type` (since I use `return_type.__name__`), but really it's a `NewType`, which is a function :/ ... so I guess it could either be a big union of all the NewTypes... anyway, you can see why I punted :joy:

File: napari/utils/_injection/_processors.py
Comment (tlambert03) [reply napari/utils/_injection/_processors.py]: thx

File: napari/utils/_injection/_accessors.py
Comment (alisterburt) [inline napari/utils/_injection/_accessors.py]: super minor, non blocking: slight naming inconsistency here, `_get_*` vs `_access_*`

**Review by alisterburt — APPROVED** [review]
I really love this, amazing work @tlambert03 ! 

I've taken another pass and nothing really jumped out at me so I'm approving, feel free to address or ignore my minor comment. 

I assume if and when the time comes that we want to open this up for external use we will define an `inject_external_dependencies` which may be called from within `inject_napari_dependencies`? Not relevant to this PR anyway, onwards!!

File: napari/utils/_injection/_accessors.py
Comment (tlambert03) [reply napari/utils/_injection/_accessors.py]: > I really love this, amazing work @tlambert03 !
> 
> I've taken another pass and nothing really jumped out at me so I'm approving, feel free to address or ignore my minor comment.
> 
> I assume if and when the time comes that we want to open this up for external use we will define an `inject_external_dependencies` which may be called from within `inject_napari_dependencies`? Not relevant to this PR anyway, onwards!!

yeah, actually... since there's nothing restricting this to napari types... perhaps we should just call it `inject_type_dependencies` or something

File: napari/utils/_tests/test_magicgui.py
Comment (andy-sweet) [inline napari/utils/_tests/test_magicgui.py]: Thanks for making the tests less sleepy!!

File: napari/utils/_tests/test_injection.py
Comment (andy-sweet) [inline napari/utils/_tests/test_injection.py:139]: Love it!

File: napari/utils/_tests/test_injection.py
Comment (andy-sweet) [inline napari/utils/_tests/test_injection.py]: Do you actually need a separate thread here? Or do you just care about the return type being a Future?

If it's the latter, and I hope it is so we can remove those little sleeps, then you could just do something like

```
future = Future()
future.set_result(np.random.rand(4, 4))
return future
```

which seems fine for testing.

In Java concurrency land, this is like calling [immediateFuture](https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/Futures.html#immediateFuture(V))

File: napari/utils/_tests/test_injection.py
Comment (andy-sweet) [inline napari/utils/_tests/test_injection.py]: Nit: I don't think we actually care about the data here, so `zeros` or `empty` (if that's viable here) should be preferred to `random` here.

File: napari/utils/_injection/_processors.py
Comment (andy-sweet) [inline napari/utils/_injection/_processors.py:187]: Is that because modifying the layerlist will emit an event that will cause a `QWidget` to be updated in non-headless napari? Maybe it should be the responsibility of the callback that's connected to the event to call `ensure_main_thread` because it is likely more aware that it is in a QApp?

File: napari/utils/_injection/__init__.py
Comment (andy-sweet) [inline napari/utils/_injection/__init__.py:41]: If I understood correctly, this decorator does more than just dependency injection as described in the docstring. It is also causes napari to process/consume the return value in some way (e.g. add a layer to the layerlist). Is that right? If so, I think it should be documented here.

After reading the rest of the PR, it might be better to rewrite the docstring in terms of accessors and processors? I think this function with find accessors to inject parameters and will also process return values. I wonder if we might want to split this into two functions (one for accessors/injection and one for processors), but that doesn't have to be in this PR.

File: napari/utils/_injection/_processors.py
Comment (andy-sweet) [inline napari/utils/_injection/_processors.py:139]: Is this actually used for anything? Looks like we just add to and remove from this set, but don't read from it. If so, remove it?

File: napari/utils/_injection/_accessors.py
Comment (andy-sweet) [inline napari/utils/_injection/_accessors.py]: I don't mind accessor, but a term I've come across in the Java world with [Dagger](https://dagger.dev/dev-guide/) and [Guice](https://github.com/google/guice/wiki/ProvidesMethods) is provider. No need to change, but just wanted to mention it in case you weren't attached to the names already.

**Review by andy-sweet — CHANGES_REQUESTED** [review]
Overall, I like this PR a lot. Some comments are question around my own understanding, but there are a few changes that I think should happen before this is merged. It's been around for a little while, already has one approver, so I'm here to add some feedback/understanding and try to get over the line.

File: napari/utils/_injection/__init__.py
Comment (tlambert03) [reply napari/utils/_injection/__init__.py:41]: > It is also causes napari to process/consume the return value in some way (e.g. add a layer to the layerlist). Is that right? If so, I think it should be documented here.

yeah, it will after this PR.  good catch.

> I wonder if we might want to split this into two functions (one for accessors/injection and one for processors), but that doesn't have to be in this PR.

I really like this.  You should be able to inject dependencies without having the output be processed (vice versa, maybe, but less important probably, since you can still override params at call time).  But, it would still be nice to have a single decorator that does both as a convenience, so do you prefer:
1. have three decorators: one for param injection, one for processors, and one for both
2. have two decorators: one for param injection with an optional `process_output: bool` argument, and one just for processors
3. other?

File: napari/utils/_injection/_accessors.py
Comment (tlambert03) [reply napari/utils/_injection/_accessors.py]: I like provider!

File: napari/utils/_injection/_processors.py
Comment (tlambert03) [reply napari/utils/_injection/_processors.py:139]: i believe it was necessary to prevent garbage collection before the future is done.  will add a comment

File: napari/utils/_injection/_processors.py
Comment (tlambert03) [reply napari/utils/_injection/_processors.py:187]: yeah, that's a good general point actually.  We could/should probably decorate a number of our viewer-updating methods with `ensure_main_thread`.  (outside of scope here)

File: napari/utils/_tests/test_injection.py
Comment (tlambert03) [reply napari/utils/_tests/test_injection.py]: removed

File: napari/utils/_injection/_processors.py
Comment (andy-sweet) [reply napari/utils/_injection/_processors.py:139]: Ah, I guess that's because normally you would create a Future with an executor (or similar), which probably does keep a reference. Explanatory comment sounds good.

File: napari/utils/_injection/__init__.py
Comment (andy-sweet) [reply napari/utils/_injection/__init__.py:41]: Given we know we want the single decorator for both, let's do that first as in this PR, then define two separate ones when needed. I dislike a boolean argument here because it adds branching behavior to the function when we could easily split the function into two.

File: napari/utils/_injection/_accessors.py
Comment (andy-sweet) [reply napari/utils/_injection/_accessors.py]: Up to you whether you change it. I'm happy either way.

**Review by andy-sweet — APPROVED** [review]

File: napari/utils/_injection/__init__.py
Comment (tlambert03) [reply napari/utils/_injection/__init__.py:41]: >I dislike a boolean argument here because it adds branching behavior to the function when we could easily split the function into two.

the bool True would simply call the second decorator as a convenience, but they would still be split into two:

```python
def inject_processor(func):
    @wraps(func)
    def _wrapped(*args, **kwargs):
        result = func(*args, **kwargs)
        process_output(result)
        return result
    return _wrapped

def inject_providers(func, process_output: bool=True):
    @wraps(func)
    def _wrapped(*args, **kwargs):
        # inject deps into args kwargs
         return func(*args, **kwargs)
    
    if process_output:
            return inject_processor(_wrapped)
    return _wrapped
```

File: napari/utils/_injection/_accessors.py
Comment (tlambert03) [reply napari/utils/_injection/_accessors.py]: I also added a couple convenience `@provider` and `@processor` decorators, similar to Guice:

```python
@provider
def _provide_int() -> int:
    ...

@processor
def _process_int(x: int) -> None:
    ...
```

File: napari/utils/_injection/_accessors.py
Comment (tlambert03) [reply napari/utils/_injection/_accessors.py]: done

File: napari/utils/_injection/__init__.py
Comment (andy-sweet) [reply napari/utils/_injection/__init__.py:41]: I mostly just have a preference for composition over branching/indentation in the spirit of one function should handle one behavior. Doesn't matter too much here as the functions here are simple and the branching logic is simple, but in general I try to avoid boolean arguments.

```python
def inject_processor(func):
    @wraps(func)
    def _wrapped(*args, **kwargs):
        result = func(*args, **kwargs)
        process(result)
        return result
    return _wrapped

def inject_providers(func):
    @wraps(func)
    def _wrapped(*args, **kwargs):
        # inject deps into args kwargs
        return func(*args, **kwargs)
    return _wrapped

def inject_providers_and_processor(func):
    return inject_processor(inject_providers(func))
```

Or if I understood correctly, the ordering of wrapping doesn't matter here, so maybe just add both decorators when needed?

```python
@inject_processor
@inject_providers
def foo(...):
    ...
```

Anyway, completely up to you which way to go. I'm approving regardless.

Comment (andy-sweet) [reply napari/utils/_injection/__init__.py:41]: Actually, one thing I like about the separate decorators is that I don't love `@inject_processor` as a name, as I don't think of what that's doing as injection. It's more like a post-processor or a done callback for a future - having separate decorators means you could call that something else unrelated to injection (e.g. `@process_output`).

Again, these are just minor comments. Definitely not blocking this PR at all.

Comment (tlambert03) [reply napari/utils/_injection/__init__.py:41]: very much agree all around.  The naming needs work (and it's not processor "injection").  Let's do in a followup.  Also not opposed to requiring the two decorators... but still holding out hope for a single decorator if one wants both

**Comment by tlambert03** [comment]
Will probably merge this tomorrow unless anyone else has comments

**Pull request merged**
